examples/fluids/problems/problem-weak.c

```diff
@@ -0,0 +1,20 @@
+#include <ceed.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <ceed/backend.h>
+#include "../navierstokes.h"
+
+// This function provides a debug target for weak symbols
+// LCOV_EXCL_START
+static PetscErrorCode ProblemRegister_Weak(const char *name) {
+  // TODO: What should be the error message?
+  if (getenv("CEED_DEBUG")) fprintf(stderr, "Weak %s\n", name);
```

Comment:
@jedbrown , I am not sure if this If-statement works fine since this function doesn't get called at all. Do I need to change the Makefile as well or is there something wrong in the implementation?

Reply:
The weak symbol will get discarded by the linker if there is a strong (normal) symbol of the same name. To test this being called, you can build without Enzyme (or otherwise skip linking the strong symbol) or just create an extra weak symbol for which there is no corresponding strong symbol. This (when you run with `CEED_DEBUG=1` in your environment) is just logging that this symbol is weak and thus will error if selected at run-time.

Reply:
I tried to build without Enzyme in the last commit which I hadn't pushed. It should be able to discard `EULER_VORTEX` when not using clang or ENZYME_LIB not provided, right?
Note I started with the fluids example since I am more familiar with that code.

Reply:
To get this to work right requires changes in the makefile and the `problems` directory.

The existing register function for all current (and new) problems should remain as is.

You also need that new file for the weak registration of any problem when the required libs are not present. It looks like you have that. It should be pretty similar to the backend and gallery weak registration functions.

In the makefile, you will want to separately add the source file for the new Enzyme 'problem' to `src.c`
https://github.com/CEED/libCEED/blob/5b1362b9c136ba2a203fe6f05a8cefc3f59aa1b3/examples/fluids/Makefile#L39
sort of like what we do when we detect that AVX is supported by the compilation environment
https://github.com/CEED/libCEED/blob/5b1362b9c136ba2a203fe6f05a8cefc3f59aa1b3/Makefile#L333

You will then need this weak_last rule to help make sure that the weak registration of problems occurs last
https://github.com/CEED/libCEED/blob/5b1362b9c136ba2a203fe6f05a8cefc3f59aa1b3/Makefile#L479

Reply:
Thanks, @jeremylt . These were very helpful.

Reply:
awesome!

Reply:
For your point 2) in the commit message, I made some changes in #826 along those lines. Each register function reports in debug mode so you can see everything that is registered. To see the effect, build the branch and run `CEED_DEBUG=1 ./build/t000-ceed /cpu/self`
You get
```
---------- Registering Backends ----------

Backend Register: /cpu/self/avx/blocked
Backend Register: /cpu/self/avx/serial
Backend Register: /gpu/cuda/ref
Backend Register: /gpu/cuda/gen
Backend Register: /gpu/cuda/shared
Backend Register: /gpu/hip/ref
Backend Register: /gpu/hip/gen
Backend Register: /gpu/hip/shared
** Weak Register: /gpu/cuda/magma
** Weak Register: /gpu/hip/magma
** Weak Register: /gpu/cuda/magma/det
** Weak Register: /gpu/hip/magma/det
Backend Register: /cpu/self/memcheck/blocked
Backend Register: /cpu/self/memcheck/serial
** Weak Register: /cpu/self/occa
** Weak Register: /cpu/openmp/occa
** Weak Register: /gpu/hip/occa
** Weak Register: /gpu/cuda/occa
Backend Register: /cpu/self/opt/blocked
Backend Register: /cpu/self/opt/serial
Backend Register: /cpu/self/ref/serial
Backend Register: /cpu/self/ref/blocked
Backend Register: /cpu/self/xsmm/blocked
Backend Register: /cpu/self/xsmm/serial
```